src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -1689,14 +1689,14 @@ fn write_fingerprint(loc: &Path, fingerprint: &Fingerprint) -> CargoResult<()> {
     // as we can use the full hash.
     let hash = fingerprint.hash_u64();
     debug!("write fingerprint ({:x}) : {}", hash, loc.display());
-    paths::write(loc, util::to_hex(hash).as_bytes())?;
+    paths::write_atomic(loc, util::to_hex(hash).as_bytes())?;
 
     let json = serde_json::to_string(fingerprint).unwrap();
     if cfg!(debug_assertions) {
         let f: Fingerprint = serde_json::from_str(&json).unwrap();
         assert_eq!(f.hash_u64(), hash);
     }
-    paths::write(&loc.with_extension("json"), json.as_bytes())?;
+    paths::write_atomic(&loc.with_extension("json"), json.as_bytes())?;
```

Comment:
(Laying out my remaining doubt from the last PR.)

Citing @epage's [comment](https://github.com/rust-lang/cargo/pull/12362#discussion_r1269787624).
> The bug is "wiped my file clean on ctrl-c" (#11386). I assume these "atomic writes" would instead be "I've got a random temp file on ctrl-c". Only slightly better. If that is the case, then I would expect this to only be applied to the cases where the empty file causes problems.

I don't know how all `cargo` components work, but I imagine fingerprints is a place where empty files are treated correctly? If so, should these be reverted?

Reply:
With the existing write, I will occasionally get zero length Cargo.toml when being destructive. With larger config, I might expect to see corrupt nonzero writes, but I have not seen this.

I consider a command that unexpectedly moves my project from a working state to non-existent to have a bug.

This change should at least keep our output config in one of 2 states, the same as before the change, or the new config content.

I think we are in danger, though small, of this bug in any place we write Cargo.toml out.  I'm also not familiar with fingerprints code, but can't imagine leaving a zero length config accidentally is not a bug.

I think we'd want to keep this change.

What am I missing?



Sorry this is.from a phone, so edits are hard.

Reply:
> What am I missing?

I believe you're missing this: the changes in question here does not write a manifest file, but a fingerprint.

(same for the change in the `src/cargo/core/compiler/build_context/target_info.rs` file)

Reply:
Thanks for hanging in there and helping me understand. 

Ok, turned it around. Only changing the `write()`s in the path for command "cargo add" and "cargo remove".  If other sections need to adopt this change, they can do so later.  

Thanks for helping me understand.